### PR TITLE
fix bugs in syncbn

### PR DESCRIPTION
**Pull request by jeffdaily**
- incorrect use of __shfl_down
- fix warp size assumptions
- update unit tests to exit on failure

### CODE REVIEW

**Comment by jeffdaily** [comment]
Also note that I have updated the CI script to run the additional distributed test script that exercises syncbn.

**Comment by jeffdaily** [comment]
The first CI run here that passed was before the CI script change; latest passing run has the change.

File: csrc/welford.cu [hunk 1]
```diff
@@ -61,25 +64,27 @@ __device__ __forceinline__ T reduce_block(T *x, T val)
...
 
   return val;
 }
 
 #define ELEMENTS_PER_ITER 4 // enables concurrency within each thread to hide latency
 #define ELEMENTS_PER_THREAD 16
-#define OPTIMAL_TILE_W 32
+#define OPTIMAL_TILE_W WARP_SIZE
 #define MAX_H_BLOCK 128
 #define MAX_BLOCK_SIZE 512
```
Comment (sarunyap) [inline csrc/welford.cu:89]: @jeffdaily Apart from the warp size fix, should we also check the launch_bounds?  Here we have MAX_BLOCK_SIZE set to 512.

File: csrc/welford.cu [hunk 2]
```diff
@@ -291,8 +292,8 @@ __global__ void welford_kernel(
     }
   }
 
-  static __shared__ int s_mem[160];
-  accscalar_t* s_mem_ac = (accscalar_t*) &s_mem[32];
+  static __shared__ int s_mem[MAX_BLOCK_SIZE/WARP_SIZE];
+  static __shared__ accscalar_t s_mem_ac[MAX_BLOCK_SIZE/WARP_SIZE*2];
```
Comment (sarunyap) [inline csrc/welford.cu:296]: Why is smem size reduced from 160 to MAX_BLOCK_SIZE/WARP_SIZE = 8 and s_mem_ac size reduced from 32 to  MAX_BLOCK_SIZE/WARP_SIZE*2 = 16? It seems like s_mem is used for storing the value from each lane (count[lane]). Should not it be of size at least 64?  s_mem_ac seems to be used for storing the value from each lane too but the offset is different (x[lane*2+1]). I think it should be of size at least 128. Did we test this kernel?

File: tests/distributed/synced_batchnorm/unit_test.sh [hunk 3]
```diff
@@ -1,8 +1,8 @@
...
-python test_batchnorm1d.py
-python -m torch.distributed.launch --nproc_per_node=2 two_gpu_unit_test.py
-python -m torch.distributed.launch --nproc_per_node=2 two_gpu_unit_test.py --fp16
-python -m torch.distributed.launch --nproc_per_node=2 two_gpu_test_different_batch_size.py --apex
+python python_single_gpu_unit_test.py || exit 1
+python single_gpu_unit_test.py || exit 1
+python test_batchnorm1d.py || exit 1
+python -m torch.distributed.launch --nproc_per_node=2 two_gpu_unit_test.py || exit 1
+python -m torch.distributed.launch --nproc_per_node=2 two_gpu_unit_test.py --fp16 || exit 1
+python -m torch.distributed.launch --nproc_per_node=2 two_gpu_test_different_batch_size.py --apex || exit 1
```
Comment (sarunyap) [inline tests/distributed/synced_batchnorm/unit_test.sh:6]: Why do we need || exit 1?  Did the processes get stuck?

File: csrc/welford.cu (hunk 2 above)
Comment (jeffdaily) [reply csrc/welford.cu:296]: Good catch.  I will update.

File: tests/distributed/synced_batchnorm/unit_test.sh (hunk 3 above)
Comment (jeffdaily) [reply tests/distributed/synced_batchnorm/unit_test.sh:6]: Our CI job doesn't even run this script (yet), but I was finding when I ran it locally that the script wouldn't report a non-zero exit code even if one or more of the tests within this script were failing.  Alternatively, I might be able to add `set -ex` at the top of the script to get similar behavior.

File: csrc/welford.cu (hunk 2 above)
Comment (jeffdaily) [reply csrc/welford.cu:296]: Great catch, thank you.

File: csrc/welford.cu (hunk 1 above)
Comment (jeffdaily) [reply csrc/welford.cu:89]: Still need to figure this out.  If I do not explicitly use `__launch_bounds__`, (so default must be 256) unit tests pass.  If I specify 512, unit tests fail.  If I specify 1024, unit tests pass again.

Comment (jeffdaily) [reply csrc/welford.cu:89]: Was not able to reproduce with latest pushed commit.

File: csrc/welford.cu (hunk 2 above)
Comment (jeffdaily) [reply csrc/welford.cu]: Fixed in latest commit.

**Review by sarunyap — APPROVED** [review]
I'm OK with the changes given that the related unit tests passed.

**Pull request merged**